packages/contracts-bedrock/src/interop/CrossL2Inbox.sol

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { ISemver } from "src/universal/ISemver.sol";
+
+/// @notice Entry to post to the inbox.
+///         The postie may deliver multiple entries per mail delivery.
+/// @custom:field chain   Chain identifier.
+/// @custom:field output  Output-root of the chain.
+struct InboxEntry {
+    bytes32 chain;
+    bytes32 output;
+}
+
+/// @custom:proxied
+/// @title CrossL2Inbox
+/// @notice The CrossL2Inbox receives output-roots of any chain,
+///         and makes the output-roots available for cross-L2 proving.
+contract CrossL2Inbox is ISemver {
+    /// @notice The address that is allowed to post into the inbox.
+    /// This is temporary for Interop Milestone 0:
+    /// this will be changed to a system-only address later.
+    address internal immutable SUPERCHAIN_POSTIE;
+
+    /// @notice The collection of output roots, by chain.
+    /// chain ID => output root => bool.
+    mapping(bytes32 => mapping(bytes32 => bool)) public roots;
+
+    /// @custom:semver 0.0.1
+    string public constant version = "0.0.1";
+
+    /// @notice Initialize the inbox.
+    /// @param _superchainPostie  Address that will be allowed to deliver to the inbox.
+    constructor(address _superchainPostie) {
+        SUPERCHAIN_POSTIE = _superchainPostie;
+    }
+
+    /// @notice Getter for the SUPERCHAIN_POSTIE address.
+    function superchainPostie() external view returns (address) {
+        return SUPERCHAIN_POSTIE;
+    }
+
+    /// @notice The inbox receives mail from the postie.
```

Comment:
See this https://x.com/Omniscia_sec/status/1727356839761764417

Reply:
Do we want to make the calldata structure more complex, by grouping output-roots of the same chain ID? There's no place to share a local intermediate storage key. It should be a single SSTORE either way. There's a double lookup to two related but different calldata struct fields, but there's nothing to share their either. I'm down to optimize this function with assembly and other tricks later FTW, but we're only in the prototype stage.

Reply:
Also, depending on the design, we might be able to include output-roots without their chain ID, by including the chain-identity in the output-root pre-image itself.

Reply:
Got it. Was just commenting on the declaring a local storage mapping when doing the nested accesses to save some hashing. I thought it was a cool pattern to share

Reply:
Putting the chain identity in the output does seem like an interesting idea but unclear what UX implications that would have. Ideally there is p2p network for serving output root preimages